[Feature] OTEL tracing during loading

emricksini-h opened the pull request:
## Purpose

This PR instruments the vLLM loading process with OpenTelemetry (OTel) tracing to provide precise, span-based observability into initialization performance.

## Why we need it
Currently, performance insights for vLLM's startup phase rely on parsing unstructured logs. Tools like [llm-d-benchmark](https://github.com/llm-d/llm-d-benchmark) depend on specific log strings to calculate durations [see here](https://github.com/llm-d/llm-d-benchmark/blob/main/workload/harnesses/nop-llm-d-benchmark.py), which introduces several limitations:
- Metrics pipelines break whenever log messages or formats are updated
- Durations are computed using timers with start/end placed arbitrarily without always reflecting the real duration of given functions
- Some parts of the loading process are currently unlogged, creating some blind spots in the initialization timeline where time is lost but cannot be visualized.

Tracing improves all of the previously mentioned limitations by allowing to track main methods duration directly from traces informations.

#### Example tracing output (Qwen3-VL-30B-A3B-Instruct-FP8)
<img width="1345" height="310" alt="Capture d’écran 2025-12-22 à 17 38 39" src="https://github.com/user-attachments/assets/3898b56b-36fa-4a07-902e-1d55e09d0f26" />

---

> [!NOTE]
> Introduces OpenTelemetry tracing for core startup and compilation paths, enabling span-level visibility and distributed context propagation.
> 
> - New `vllm/tracing.py` provides `init_tracer`, `instrument` decorator, trace header extraction/propagation, worker tracer init, and OTLP exporters
> - Instruments key phases: `Load model`, `Load weights`, `Initialize model`, CPU/GPU `Loading`/`Warmup`/`Capture model`, KV cache init/allocation, executor/worker/client inits, API server setup, and compile spans (`Compile graph`, `Inductor compilation`); records Dynamo transform time when available
> - Adds optional `extras_require["otel"]` in `setup.py` for OTel packages
> - Minor refactor: extracted `_run_inductor_compilation` in backends and replaced direct interpreter call
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 9694af5ca20b0263ade5df6417ec5856437334ec. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Adds span-based tracing across initialization and compilation with optional OTel deps.
> 
> - **New tracing runtime**: `vllm/tracing.py` with `init_tracer`, `instrument` decorator, trace header extraction/propagation, worker tracer init, OTLP exporters, and env-based context propagation
> - **Instrumentation**: spans added to API server setup, executor/worker/client inits, model init/loading/weight download, CPU/GPU loading/warmup/capture, KV cache prep/allocation, compilation phases (`Compile graph`, `Inductor compilation`), and Dynamo transform time recorded when available
> - **Build/Deps**: new `extras_require["otel"]` in `setup.py`
> - **Refactor**: extracted `_run_inductor_compilation` in `compilation/backends.py` and replaced direct interpreter call
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ba2c300124a085ce4cb3f9d38dcc7f067455dfbd. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Adds span-based observability to startup and compile paths via OpenTelemetry, with optional install and distributed context propagation.
> 
> - New `vllm/tracing.py`: `init_tracer`, `instrument` decorator, OTLP exporters (gRPC/HTTP), env/header context extract/propagate, `maybe_init_worker_tracer`
> - Instrumented key phases with spans: `API server setup`, `Executor/Worker/Client` inits, `Initialize model`/`Load model`/`Load weights`, CPU/GPU `Loading`/`Warmup`/`Capture model`, `Prepare model`, `Allocate KV cache`, `DeepGemm warmup`, `Compile graph` and `Inductor compilation`; records Dynamo bytecode transform time when OTel is available
> - Optional deps: `extras_require["otel"]` in `setup.py`
> - Minor refactor: extract `_run_inductor_compilation` in `compilation/backends.py` and replace direct call
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 831d555a7fdc765baa4075ff080e587c067cea37. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Adds span-based observability with OpenTelemetry and distributes trace context across processes.
> 
> - New `vllm/tracing.py`: `init_tracer`, `instrument` decorator, env/header context extraction/propagation, `maybe_init_worker_tracer`, OTLP exporters (gRPC/HTTP)
> - Instruments key phases: `API server setup`, executor/worker/client init, `Initialize model`/`Load model`/`Load weights`, `Download weights - HF`, CPU/GPU `Loading`/`Warmup`/`Capture model`, `Prepare model`, `Allocate KV cache`, `DeepGemm warmup`, `Compile graph`/`Inductor compilation`; records Dynamo transform time as span when available
> - Adds optional deps `extras_require["otel"]` in `setup.py`
> - Minor refactor: extracted `_run_inductor_compilation` and used it in place of direct interpreter call
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit bc0c0cea3cb4866d68f733ed0d66c40793ed317e. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

emricksini-h committed:
Enhance tracing init for multi-process + add instrument decorator

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

emricksini-h committed:
Add tracing to main loading methods

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

emricksini-h committed:
Fix double import

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

emricksini-h committed:
Put OTEL dependencies in extra_require

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

emricksini-h committed:
fix pre-commit

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

emricksini-h committed:
Some fixes

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

aarnphm commented:
I'm in favor of adding tracing in general, as an opt-in, but will probably need some coordination to add tracing to multiple components within vLLM here, therefore I would encourage to have a RFC for tracing in general then substrate it to smaller component (like you have here for startup)

benchislett reviewed (commented):
Could you move `tracing.py` to something like `otel_tracing.py` and then have a generic `def instrument` in `tracing.py`? I really like how non-invasive the instrumentation is and I think we will want (at some point) to hook other backends (profilers, etc) into the instrumentation function. For now it can only dispatch to otel, but I think a high-value and low-effort change would be to make the foremost function a bit more flexible, and not explicitly linked to OpenTelemetry.

emricksini-h committed:
fix dynamo start time error

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

emricksini-h committed:
add generic tracing layer

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

gcanlin commented:
@emricksini-h Hi! I would like to install OTEL with this PR. But I met some installation problems. Could you please help take a look?

```
git fetch pull/31162/head:pr-31162
git checkout pr-31162
 pip install -v ".[otel]"
```

```
    File "/home/guocanlin/vllm-workspace/vllm/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 70, in _build_backend
      obj = import_module(mod_path)
            ^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/guocanlin/.local/share/uv/python/cpython-3.12.3-linux-x86_64-gnu/lib/python3.12/importlib/__init__.py", line 90, in import_module
      return _bootstrap._gcd_import(name[level:], package, level)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
    File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
    File "<frozen importlib._bootstrap>", line 1310, in _find_and_load_unlocked
    File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
    File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
    File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
    File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
    File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
    File "<frozen importlib._bootstrap_external>", line 995, in exec_module
    File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
    File "/tmp/pip-build-env-089ldckd/overlay/lib/python3.12/site-packages/setuptools/__init__.py", line 16, in <module>
      import setuptools.version
    File "/tmp/pip-build-env-089ldckd/overlay/lib/python3.12/site-packages/setuptools/version.py", line 1, in <module>
      import pkg_resources
    File "/tmp/pip-build-env-089ldckd/overlay/lib/python3.12/site-packages/pkg_resources/__init__.py", line 2172, in <module>
      register_finder(pkgutil.ImpImporter, find_on_path)
                      ^^^^^^^^^^^^^^^^^^^
  AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> No available output.

  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /home/guocanlin/vllm-workspace/vllm/.venv/bin/python /home/guocanlin/vllm-workspace/vllm/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py get_requires_for_build_wheel /tmp/tmpvqvi6o6k
  cwd: /tmp/pip-install-c956nqw8/numpy_289b9308ac9845bc899e2c7c94855bb8
  Getting requirements to build wheel ... error
ERROR: Failed to build 'numpy' when getting requirements to build wheel
```

emricksini-h committed:
allow more recent otel deps

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

gcanlin commented:
@emricksini-h Hey, thank you for the quick reply! I try the newest code and also get the error below. But I manually execute `pip install opentelemetry-sdk>=1.26.0 opentelemetry-api>=1.26.0 opentelemetry-exporter-otlp>=1.26.0 opentelemetry-semantic-conventions-ai>=0.4.1` and I install them successfully. I guess the error caused for I have built vLLM. Not sure that. I prefer to think it's irrelevant about this PR.

BTW, I'm stuck at how to enable OTEL tracing for vLLM and get the trace view. Could you give some examples or guidance? Thanks!

```
opentelemetry-api                        1.39.0
opentelemetry-exporter-otlp              1.39.0
opentelemetry-exporter-otlp-proto-common 1.39.0
opentelemetry-exporter-otlp-proto-grpc   1.39.0
opentelemetry-exporter-otlp-proto-http   1.39.0
opentelemetry-proto                      1.39.0
opentelemetry-sdk                        1.39.0
opentelemetry-semantic-conventions       0.60b0
opentelemetry-semantic-conventions-ai    0.4.13
```


```
      cmd_obj.run()
    File "/tmp/pip-build-env-11my_vvd/overlay/lib/python3.12/site-packages/setuptools/_distutils/command/build.py", line 135, in run
      self.run_command(cmd_name)
    File "/tmp/pip-build-env-11my_vvd/overlay/lib/python3.12/site-packages/setuptools/_distutils/cmd.py", line 357, in run_command
      self.distribution.run_command(command)
    File "/tmp/pip-build-env-11my_vvd/overlay/lib/python3.12/site-packages/setuptools/dist.py", line 1102, in run_command
      super().run_command(command)
    File "/tmp/pip-build-env-11my_vvd/overlay/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 1021, in run_command
      cmd_obj.run()
    File "<string>", line 360, in run
    File "/tmp/pip-build-env-11my_vvd/overlay/lib/python3.12/site-packages/setuptools/command/build_ext.py", line 96, in run
      _build_ext.run(self)
    File "/tmp/pip-build-env-11my_vvd/overlay/lib/python3.12/site-packages/setuptools/_distutils/command/build_ext.py", line 368, in run
      self.build_extensions()
    File "<string>", line 317, in build_extensions
    File "<string>", line 294, in configure
    File "/home/guocanlin/.local/share/uv/python/cpython-3.12.3-linux-x86_64-gnu/lib/python3.12/subprocess.py", line 413, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['cmake', '/home/guocanlin/vllm-workspace/vllm', '-G', 'Ninja', '-DCMAKE_BUILD_TYPE=RelWithDebInfo', '-DVLLM_TARGET_DEVICE=cuda', '-DVLLM_PYTHON_EXECUTABLE=/home/guocanlin/vllm-workspace/vllm/.venv/bin/python', '-DVLLM_PYTHON_PATH=/tmp/pip-build-env-11my_vvd/site:/home/guocanlin/.local/share/uv/python/cpython-3.12.3-linux-x86_64-gnu/lib/python312.zip:/home/guocanlin/.local/share/uv/python/cpython-3.12.3-linux-x86_64-gnu/lib/python3.12:/home/guocanlin/.local/share/uv/python/cpython-3.12.3-linux-x86_64-gnu/lib/python3.12/lib-dynload:__editable__.vllm-0.11.2.dev698+gcf4a69da1.d20251213.precompiled.finder.__path_hook__:/tmp/pip-build-env-11my_vvd/overlay/lib/python3.12/site-packages:/tmp/pip-build-env-11my_vvd/normal/lib/python3.12/site-packages:/tmp/pip-build-env-11my_vvd/overlay/lib/python3.12/site-packages/setuptools/_vendor:/tmp/pip-build-env-11my_vvd/overlay/lib/python3.12/site-packages/grpc_tools/_proto', '-DFETCHCONTENT_BASE_DIR=/home/guocanlin/vllm-workspace/vllm/.deps', '-DNVCC_THREADS=1', '-DCMAKE_JOB_POOL_COMPILE:STRING=compile', '-DCMAKE_JOB_POOLS:STRING=compile=128', '-DCMAKE_CUDA_COMPILER=/usr/local/cuda/bin/nvcc']' returned non-zero exit status 1.
  error: subprocess-exited-with-error

  × Building wheel for vllm (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> No available output.

  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /home/guocanlin/vllm-workspace/vllm/.venv/bin/python /home/guocanlin/vllm-workspace/vllm/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py build_wheel /tmp/tmpa_02d29p
  cwd: /home/guocanlin/vllm-workspace/vllm
  Building wheel for vllm (pyproject.toml) ... error
  ERROR: Failed building wheel for vllm
Failed to build vllm
error: failed-wheel-build-for-install

× Failed to build installable wheels for some pyproject.toml based projects
╰─> vllm
```

simon-mo reviewed (commented):
* Please add a test in test tracing? 
* Should be consider unify the name style to be `module_name.class_name::method_name`?

emricksini-h committed:
Changes following review

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

emricksini-h committed:
Add tracing tests

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

emricksini-h commented:
Thanks for the review @simon-mo.

> Please add a test in test tracing?

Done. I have added them to `tests/tracing` since this functionality isn't restricted to V1. I moved the reusable components from `tests/v1/test_tracing.py` (such as `trace_service`) into `tests/tracing/conftest.py`.

> Should be consider unify the name style to be module_name.class_name::method_name?

Are you referring to the span names? If so, I think keeping the ability to define custom names is valuable. It allows us to use names that are more representative of the specific step occurring, which makes traces easier to understand from a high-level perspective.

simon-mo commented:
FYI @ProExpertProg @zou3519 for measurement of startup time UX

emricksini-h committed:
Better handling when OTel not available

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>

emricksini-h committed:
Fix tests

Signed-off-by: emricksini-h <emrick.birivoutin@hcompany.ai>
